prboom2/src/gl_preprocess.c

```diff
@@ -695,6 +706,83 @@ static void gld_PrecalculateSector(int num)
   Z_Free(lineadded);
 }
 
+static dboolean gld_SectorIsSelfReferencing(sector_t* sector)
+{
+  int i;
+
+  for (i = 0; i < sector->linecount; ++i)
+  {
+    line_t* l = sector->lines[i];
+    if (l->sidenum[0] != NO_INDEX &&
+        l->sidenum[1] != NO_INDEX &&
+        sides[l->sidenum[0]].sector == sector &&
+        sides[l->sidenum[1]].sector == sector)
+        return true;
+  }
+  return false;
+}
+
+static sector_t* gld_SelfReferencingSectorContainer(sector_t* sector)
+{
+  int i;
+  dboolean selfref = false;
+
+  for (i = 0; i < sector->linecount; ++i)
+  {
+    line_t* l = sector->lines[i];
+    if (l->sidenum[0] != NO_INDEX &&
+        l->sidenum[1] != NO_INDEX &&
+        sides[l->sidenum[0]].sector == sector &&
+        sides[l->sidenum[1]].sector == sector)
+    {
+      // Found a self-referencing line, look for sectors on either side of it
+      // by projecting a fraction of a unit out orthogonally from the midpoint.
+      angle_t ang = R_PointToAngle2(l->v1->x, l->v1->y, l->v2->x, l->v2->y) + ANG90;
+      fixed_t offsx = finecosine[ang >> ANGLETOFINESHIFT] >> 5;
+      fixed_t offsy = finesine[ang >> ANGLETOFINESHIFT] >> 5;
+      fixed_t xmid = (l->v1->x + l->v2->x) >> 1;
+      fixed_t ymid = (l->v1->y + l->v2->y) >> 1;
+      fixed_t xp1 = xmid + offsx;
+      fixed_t yp1 = ymid + offsy;
+      fixed_t xp2 = xmid - offsx;
+      fixed_t yp2 = ymid - offsy;
+      sector_t* s1 = R_PointInSubsector(xp1, yp1)->sector;
+      sector_t* s2 = R_PointInSubsector(xp2, yp2)->sector;
+
+      selfref = true;
+
+      if (s1 != NULL && s1 != sector)
+        return s1;
+      if (s2 != NULL && s2 != sector)
+        return s2;
+    }
+  }
+
+  if (selfref)
+  {
+    // We're self-referencing, but we couldn't find another sector across
```

Comment:
I think this part of the logic is susceptible to false positives, so I'm going to rework it